### DISCUSSION TITLE
Spring Batch 5 with MySQL Replication

### DISCUSSION THREAD

**Discussion by valliappanmvi** [spring-projects/spring-batch#4638] [Q&A] [answered]
I am using Spring Batch-5 and I see the MYSQL for BATCH5_JOB_EXECUTION_PARAMS doesn't have the primary key defined.

During runtime, I face the exception

Exception Occurred when Loading Mdynamics Configuration -> org.springframework.jdbc.UncategorizedSQLException: PreparedStatementCallback; uncategorized SQLException for SQL
[INSERT INTO BATCH5_JOB_EXECUTION_PARAMS(JOB_EXECUTION_ID, PARAMETER_NAME, PARAMETER_TYPE, PARAMETER_VALUE, IDENTIFYING)

	VALUES (?, ?, ?, ?, ?)
]; SQL state [HY000]; error code [3098]; The table does not comply with the requirements by an external plugin.

Is any way to overcome this ?

**Comment by fmbenhassine** [comment]
> The table does not comply with the requirements by an external plugin.

I am curious to know more about this external plugin. Can you share its name?

The table does not have a primary key because records represent value objects. But this does not prevent you from defining a primary key, and should not impact the framework's internals in any way. Have you tried that?

**Reply by valliappanmvi** [reply]
The error message "SQL state [HY000]; error code [3098]; The table does not comply with the requirements by an external plugin." is thrown by the MySQL driver.

**Reply by fmbenhassine** [reply, marked as answer]
Thank you for the link! Indeed, it is a best practice to define a primary key. 

> We can create primary key for the table, but was wondering how the table doesn't cater for such requirement.

As mentioned above, that table does not define a primary key because records represent value objects. But you can define one when you use MySQL replication.

**Comment by valliappanmvi** [comment]
Hi, Mahmoud

Came across the below,

https://dev.mysql.com/doc/refman/8.4/en/group-replication-requirements.html#:~:text=Every%20table%20that%20is%20to,a%20non%2Dnull%20unique%20key<https://dev.mysql.com/doc/refman/8.4/en/group-replication-requirements.html#:~:text=Every%20table%20that%20is%20to,a%20non-null%20unique%20key>.

***@***.***

As per the MySQL docs, all the table must have primary key.

We can create primary key for the table, but was wondering how the table doesn't cater for such requirement.

Thanks,
Valliappan. S
